ml_perf/reference_implementation.py

```diff
@@ -0,0 +1,279 @@
+  logging.info('Running %s:\n  %s', name, '\n  '.join(cmd))
+  with utils.logged_timer('%s finished' % name.capitalize()):
+    completed_process = subprocess.run(
+        cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+    if completed_process.returncode:
+      logging.error('Error running %s: %s', name,
+                    completed_process.stdout.decode())
+      raise RuntimeError('Non-zero return code executing %s' % ' '.join(cmd))
+  return completed_process
+
+
+def get_lines(completed_process, slice):
+  return '\n'.join(completed_process.stdout.decode()[:-1].split('\n')[slice])
+
+
+class MakeSlice(object):
+
+  def __getitem__(self, item):
+    return item
+
+
+make_slice = MakeSlice()
+
+
+def cc_flags(state):
+  return [
+      '--engine={}'.format(FLAGS.engine),
+      '--virtual_losses=8',
+      '--seed={}'.format(state.seed),
+  ]
+
+
+def py_flags(state):
+  return [
+      '--work_dir={}'.format(fsdb.working_dir()),
+      '--trunk_layers=10',
+      '--conv_width=64',
+      '--value_cost_weight=0.25',
+      '--training_seed={}'.format(state.seed),
+  ]
+
+
+# Generate an initial model with random weights.
+def bootstrap(state):
+  checked_run([
+      'python3', 'bootstrap.py', '--export_path={}'.format(
+          state.play_model_path)
+  ] + py_flags(state), 'bootstrap')
+
+
+# Freezes a model checkpoint into a TensorFlow GraphDef proto, which is the
+# required model format for C++ TensorFlow.
+def freeze(state, model_path):
+  checked_run([
+      'python3', 'freeze_graph.py', '--model_path={}'.format(model_path)
+  ] + py_flags(state), 'freeze')
+
+
+# Self-play a number of games.
+def selfplay(state):
+  play_output_name = state.play_output_name
+  play_output_dir = os.path.join(fsdb.selfplay_dir(), play_output_name)
+  play_holdout_dir = os.path.join(fsdb.holdout_dir(), play_output_name)
+
+  result = checked_run([
+      'bazel-bin/cc/selfplay', '--parallel_games=2048',
+      '--num_readouts=100', '--model={}.pb'.format(
+          state.play_model_path), '--output_dir={}'.format(play_output_dir),
+      '--holdout_dir={}'.format(play_holdout_dir)
+  ] + cc_flags(state), 'selfplay')
+  logging.info(get_lines(result, make_slice[-2:]))
+
+  # Write examples to a single record.
+  pattern = os.path.join(play_output_dir, '*', '*.zz')
+  logging.info('Extracting examples from "{}"'.format(pattern))
+  random.seed(state.seed)
+  tensorflow.set_random_seed(state.seed)
+  numpy.random.seed(state.seed)
+  buffer = example_buffer.ExampleBuffer(sampling_frac=1.0)
+  buffer.parallel_fill(tensorflow.gfile.Glob(pattern))
+  buffer.flush(
+      os.path.join(fsdb.golden_chunk_dir(), play_output_name + '.tfrecord.zz'))
+
+
+# Train a new model.
+def train(state, tf_records):
+  result = checked_run([
+      'python3',
+      'train.py',
+      *tf_records,
+      '--export_path={}'.format(state.train_model_path),
+  ] + py_flags(state), 'training')
+  logging.info(get_lines(result, make_slice[-8:-8]))
```

Comment:
is -8:-8 intentional? that's a slice of size 0.

Reply:
lol, probably not. I've removed the first -8 for now. I'll figure the correct numbers once I have the the RL loop running for real.